### PR TITLE
HPCC-26139 Do not treat kubectl stderr output as error

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Tested case where job had finished before scheduled checked. PodScheduled=True remains after job has finished, meaning [correctly] the pending timeout mechanism is not triggers, it loops and discovers job complete via checkJobExitCode.

### CODE REVIEW

**Comment by jakesmith** [comment]
@richardkchapman - commoned the executions of kubectl so output/error consistent, an error code from the command is always treated as an exception, but output on stderr is not.

File: common/workunit/workunit.cpp [hunk 1]
```diff
@@ -14137,23 +14137,31 @@ KeepK8sJobs translateKeepJobs(const char *keepJob)
...
 
+// NB: will fire an exception if command fails (returns non-zero exit code)
+static void runKubectlCommand(StringBuffer *output, const char *title, const char *cmd, const char *input, bool logOutput, bool logErrors)
+{
+    StringBuffer _output, error;
+    if (!output)
+        output = &_output;
+    bool ret = runExternalCommand(title, *output, error, cmd, input);
+    if (logOutput && output->length())
+        DBGLOG("%s: ret=%u, stdout=%s", cmd, ret, output->trimRight().str());
```
Comment (richardkchapman) [inline common/workunit/workunit.cpp:14148]: If output contains \n this should really be split into multiple DBGLOG calls (I think there's a function to do so somewhere but I can't remember where/what it is called)

File: common/workunit/workunit.cpp [hunk 2]
```diff
@@ -14137,23 +14137,31 @@ KeepK8sJobs translateKeepJobs(const char *keepJob)
...
+    if (logErrors && error.length())
+        DBGLOG("%s: ret=%u, stderr=%s", cmd, ret, error.trimRight().str());
```
Comment (richardkchapman) [inline common/workunit/workunit.cpp:14150]: Similarly, multiline support

File: common/workunit/workunit.cpp [hunk 3]
```diff
@@ -14137,23 +14137,31 @@ KeepK8sJobs translateKeepJobs(const char *keepJob)
...
 }
 
+// NB: will fire an exception if command fails (returns non-zero exit code)
+static void runKubectlCommand(StringBuffer *output, const char *title, const char *cmd, const char *input, bool logOutput, bool logErrors)
+{
+    StringBuffer _output, error;
+    if (!output)
+        output = &_output;
+    bool ret = runExternalCommand(title, *output, error, cmd, input);
+    if (logOutput && output->length())
```
Comment (richardkchapman) [inline common/workunit/workunit.cpp:14147]: I wonder if should be (logOutput || traceLevel > 5) or similar...

File: common/workunit/workunit.cpp [hunk 4]
```diff
@@ -14271,19 +14271,7 @@ bool applyK8sYaml(const char *componentName, const char *wuid, const char *job,
...
-        DBGLOG("kubectl error: %s", error.trimRight().str());
-    if (ret)
-    {
-        DBGLOG("Using yaml %s", jobYaml.str());
-        StringBuffer errorText("Failed to replace k8s resource");
-        if (error.length())
-            errorText.append(", error: ").append(error);
-        throw makeStringException(0, errorText);
-    }
+    runKubectlCommand(nullptr, componentName, "kubectl replace --force -f -", jobYaml, true, true);
```
Comment (richardkchapman) [inline common/workunit/workunit.cpp:14274]: I know the old code did, but I'm not convinced that unconditional logging of output here is wise.

File: common/workunit/workunit.cpp [hunk 5]
```diff
@@ -14137,23 +14137,31 @@ KeepK8sJobs translateKeepJobs(const char *keepJob)
     return KeepK8sJobs::none;
 }
 
+// NB: will fire an exception if command fails (returns non-zero exit code)
+static void runKubectlCommand(StringBuffer *output, const char *title, const char *cmd, const char *input, bool logOutput, bool logErrors)
```
Comment (richardkchapman) [inline common/workunit/workunit.cpp:14141]: logErrors is always true - is it worth having?

**Review by richardkchapman — APPROVED** [review]
Some minor comments. Looks ok though

File: common/workunit/workunit.cpp (hunk 5 above)
Comment (jakesmith) [reply common/workunit/workunit.cpp:14141]: probably not, I'll remove

File: common/workunit/workunit.cpp (hunk 3 above)
Comment (jakesmith) [reply common/workunit/workunit.cpp:14147]: could just leave it down to jlog log levels.
It was recently reduced to 80 - reducing any lower would prevent DBGLOG /debugInfo logging from being issued.
I don't think we can do that yet, until all DBGLOG logging is looked at, there's bound to be some that should really be at a higher level.

But I think we can change this logging here to use a higher (than default) level, so it is not output unless component jlog level is raised, i.e. ExtraneousMsgThreshold

File: common/workunit/workunit.cpp (hunk 1 above)
Comment (jakesmith) [reply common/workunit/workunit.cpp:14148]: I've looked, but can't find such a function.
I can add one to jlib (I think you'd also want to use it https://github.com/hpcc-systems/HPCC-Platform/pull/15130 ?)

**Comment by jakesmith** [comment]
@richardkchapman - please see new commit.
Letting jlog levels decide whether to log, with errors at MCinternalError level, and output at MCExtraneousInfo (verbose).
So normally there will be no output logged, unless someone raises the logging level in a component or globally.

**Review by ghalliday — DISMISSED** [review]
@jakesmith generally much cleaner.  One change that should probably be done.

File: common/workunit/workunit.cpp [hunk 6]
```diff
@@ -14137,23 +14137,31 @@ KeepK8sJobs translateKeepJobs(const char *keepJob)
     return KeepK8sJobs::none;
 }
 
+// NB: will fire an exception if command fails (returns non-zero exit code)
+static void runKubectlCommand(StringBuffer *output, const char *title, const char *cmd, const char *input)
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply common/workunit/workunit.cpp]: have moved output to end.

File: common/workunit/workunit.cpp [hunk 7]
```diff
@@ -14137,23 +14137,31 @@ KeepK8sJobs translateKeepJobs(const char *keepJob)
...
+{
+    StringBuffer _output, error;
+    if (!output)
+        output = &_output;
+    bool ret = runExternalCommand(title, *output, error, cmd, input);
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply common/workunit/workunit.cpp]: whoops, yes, changed.

**Comment by jakesmith** [comment]
@ghalliday - please see last commit.

**Review by ghalliday — APPROVED** [review]
please squash.

**Comment by jakesmith** [comment]
@ghalliday - squashed

**Pull request merged**